src/mapgen.cpp

```diff
@@ -859,39 +860,28 @@ class jmapgen_loot : public jmapgen_piece {
             if( magazine < 0 || magazine > 100 ) {
                 jsi.throw_error( "magazine chance out of range", "magazine" );
             }
+
+            // All the probabilities are 100 because we do the roll in @ref apply.
+            Item_group *result = new Item_group( Item_group::Type::G_COLLECTION, 100, ammo, magazine );
+            if( group.empty() ) {
+                result->add_item_entry( name, 100 );
+            } else {
+                result->add_group_entry( group, 100 );
+            }
+            result_group.reset( result );
         }
 
         void apply( map &m, const jmapgen_int &x, const jmapgen_int &y, const float /*mon_density*/ ) const override
         {
             if( rng( 0, 99 ) < chance ) {
-                std::vector<item> spawn;
-                if( group.empty() ) {
-                    spawn.emplace_back( name, calendar::turn );
-                } else {
-                    spawn = item_group::items_from( group, calendar::turn );
-                }
-
-                for( auto &e: spawn ) {
-                    bool spawn_ammo = rng( 0, 99 ) < ammo && e.ammo_remaining() == 0;
-                    bool spawn_mag  = rng( 0, 99 ) < magazine && !e.magazine_integral() && !e.magazine_current();
-
-                    if( spawn_mag || spawn_ammo ) {
-                        e.contents.emplace_back( e.magazine_default(), e.bday );
-                    }
-                    if( spawn_ammo ) {
-                        e.ammo_set( default_ammo( e.ammo_type() ) );
-                    }
-                }
+                const std::vector<item> spawn = result_group->create( calendar::turn );
                 m.spawn_items( tripoint( rng( x.val, x.valmax ), rng( y.val, y.valmax ), m.get_abs_sub().z ), spawn );
             }
         }
 
     private:
-        const std::string group;
-        const std::string name;
+        std::unique_ptr<Item_spawn_data> result_group;
```

Comment:
You probably don't need the `unique_ptr` anymore. There are no polymorphic classes.